Procedures API

elenaf9 opened the pull request:
# Description of change

This PR is drafting a new API for Strongholds procedures. The goal is to add more procedures to Stronghold, based on the cryptographic primitives in [crypto.rs](https://github.com/iotaledger/crypto.rs), and to allow the user to create complex procedures that consist of multiple steps. This is especially useful for the execution of procedures on a remote Stronghold, since a chain of multiple procedures like 
```
Slip10Generate -> Slip10Derive -> Ed25519Sign
```
can then be executed through a single request.

###  Concept

_See [this HackMd](https://hackmd.io/-LIn984FQ0CmZroSiBz5mg) for detailed description._

Procedures are classified into 4 different types, depending on whether they use an existing secret and/ or create a new secret that is written into the Vault:
- `GenerateSecret`: no existing secret used, new secret created. e.g. `Slip10Generate`
- `DeriveSecret`: use existing secret, create new secret. e.g. `Slip10Derive`
- `UseSecret`: use existing secret, no new secret created. e.g. `Eded5519Sign`
- `ProcessOutput`: no secret used, no secret created. e.g. `SHA256Digest`

Apart from that, Procedures may have plain input and output, e.g.the input message & signed output in case of `Ed25519Sign`.
The output of each Procedure (if there is an output) is written into a HashMap (=`CollectedOutput`), that is passed to the subsequent procedures and returned to the user in the end. For each Procedure that has plain input, it may be set dynamically whether the procedure should have a pre-set input, or if it gathers its input from the `CollectedOutput` (= uses the output of a preceding procedure). Analogous to this, `DeriveSecret`-/`UseSecret`-Procedures may also use either existing secrets, or secrets that are created in a preceding procedure-step during execution.
Products of a procedure (new secret /  output data) are only temporary if not specified otherwise via `Procedure::write_secret(..)`/ `Procedure::store_output(..)`. Temporary secrets (= records in the vault) are revoked when the procedure returns, and temporary output data is removed from the `CollectedOutput` before returning. 
In the below code example, the seed created in `Slip10Generate` would only be temporary, whereas for ` Slip10Derive ` the new private key would stay in the vault and the output chain code is written into the `CollectedOutput` and returned to the user.
```rust
let sh = Stronghold::init_stronghold_system("my-client".into(), vec![]).await?;

// Create a new seed, that only temporary exists during execution time
let generate = Slip10Generate::default();

let chain = Chain::from_u32(vec![0]);
// Use the newly created seed as input
let derive = Slip10Derive::new_from_seed(generate.target(), chain);

// Identifier / Key for a procedure's output
let k = OutputKey::new("chain-code");

// Target vault + record, in which the derived key will be written
let private_key = Location::generic("my-vault", "child-key");

let hint = RecordHint::new("private-key").unwrap();
let derive = derive
     .store_output(k) // Configure that the output should be returned after execution
     .write_secret(private_key, hint); // Set the record to be permanent in the vault

// Execute the procedures
let output = sh.runtime_exec(generate.then(derive)).await??;

// Get the desired output
let chain_code: ChainCode = output.single_output().unwrap();
```
__Note:__ Procedures are treated as _atomic_: if a single procedure-step in the chain errors, all newly created records are revoked and garbage collected, the collected output is dropped and the error is returned.

**Crypto.rs:**
For the cryptographic primitives some traits/ structs have been added to abstract from the concrete implementations. For now, they are rather basic and only part of the procedures API. However, it can be discussed whether some of these changes should be also added to crypto.rs, to have a unified interface for algorithms of the same type (at the current point in time, crypto.rs only defines traits for the Aead algorithms). 
Some first ideas are written in [this HackMd](https://hackmd.io/X5OCQIGdS9e6Eq9cpCnO1g).

## Type of change

- [X] Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Added new procedures tests.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

elenaf9 committed:
feat(procs): brainstorm traits for proc exec

elenaf9 committed:
feat(procs): brainstorm traits for proc builder

elenaf9 committed:
feat(procs): Api Draft for creating & chaining Procs

elenaf9 committed:
feat(procs): clean types & generics

elenaf9 committed:
feat(procs): improve combine-procedures logic

elenaf9 committed:
feat(procs): add existing procedures

elenaf9 committed:
feat(procs): macros for primitive procedures

elenaf9 committed:
feat(proc): allow dynamic procedure input

elenaf9 committed:
feat(procs): (re-)add tests for complex procs

elenaf9 committed:
feat(procs): enhance API, add tests, bugfix

elenaf9 committed:
feat(procs): enhance `Procedure` derive-macro

elenaf9 committed:
feat(procs): Config of Procedure Input & Output

Enforce explicit configuration of Input- and Output-Data for a
Procedure, instead of always using the preceding procedure's output as
input for the next one.
Change the return value of a Procedure to be a HashMap, that may also
contains output data of intermediate procedure-steps
Allow output to the vault to be only temporary, in which case the record
is revoked after the procedure finished.

elenaf9 committed:
refactor(procs): rename types, clean modules

elenaf9 committed:
feat(proc): Enable remote Execution of Procedures

Implement `Serialize` and `Deserialize` for all types in the procedure
module, to allow execution on a remote peer.
Add wrapper type `PrimitiveProcedure` that wraps all possible primitive
Procedures. `ProcedureStep::then` now collects PrimitiveProcedures in a
Vec, which are sequentially executed when running the Procedure.
Procedure does not need a generic Parameter anymore.

elenaf9 committed:
Merge branch 'dev' of github.com:iotaledger/stronghold.rs into feat/procedure-api

elenaf9 committed:
fix(procs): fix tests and randomness

elenaf9 committed:
fix(procs): remove old procedures module

elenaf9 committed:
feat(procs): add hash, hmac

elenaf9 committed:
feat(procs): abstract type of hash/hmac; add aead

elenaf9 committed:
feat(procs): Add abstraction for Keys & Signatures

felsweg-iota commented on client/src/interface.rs:
```diff
@@ -822,19 +823,18 @@ impl Stronghold {
-        match result {
-            Ok(ok) => ResultMessage::Ok(ok),
-            Err(err) => ResultMessage::Error(err.to_string()),
+        match unwrap_result_msg!(actor.send(send_request).await) {
```
Good idea with macro, but what happens when a [MailboxError](https://docs.rs/actix/0.12.0/actix/enum.MailboxError.html) occurs?

elenaf9 replied:
It would `return anyhow::anyhow!(mailbox_error)`. But I'll add a proper error type for this, instead of using anyhow.

tensor-programming replied:
I concur.  A real error type would be useful here.

felsweg-iota commented on client/src/procedures.rs:
```diff
@@ -0,0 +1,440 @@
+        self.aggregated_output
+            .get(key)
+            .map(|(data, _)| data)
+            .ok_or_else(|| anyhow::anyhow!("Missing Data"))
```
Better use specific error types eg `ProcedureError::MissingData` or similar.

elenaf9 replied:
I agree.

felsweg-iota commented on client/src/procedures.rs:
```diff
@@ -0,0 +1,440 @@
+    }
+}
+
+impl From<ProcedureIo> for Vec<u8> {
```
The implementation for `From` for `ProcedureIO` is complemented with the automatic implementation of `Into` for the target type `Vec<u8>`, so no need here to implement it again :)

elenaf9 replied:
But `Into<ProcedureIO>` is not implemented for `Vec<u8>` (and if it was, this `impl From<ProcedureIo> for Vec<u8>` would not be possible because of conflicting implementations). 
Or what exactly do you mean?

felsweg-iota replied:
yes, true that. I have overseen the tuple field inside [`ProcedureIo`]

felsweg-iota commented on client/src/procedures.rs:
```diff
@@ -0,0 +1,440 @@
+    pub output: T,
+}
+
+trait Parse {
```
Better use nouns for traits or types, but that's minimal

elenaf9 replied:
I don't think there is a fixed convention for this; the traits of `std` are not really consistent on this, e.g.: 

- `Borrow`, `Debug`: Verbs
- `Iterator`: Noun
- `Sized`: Adjective

felsweg-iota commented on client/src/procedures.rs:
```diff
@@ -0,0 +1,440 @@
+    }
+
+    fn random_path() -> impl Into<Vec<u8>> {
+        let l = (rand::random::<u8>() % (u8::MAX - 1)) + 1;
```
Better move functions, that induce randomness, to some other module.

felsweg-iota commented on client/src/procedures.rs:
```diff
@@ -0,0 +1,440 @@
+    fn random_path() -> impl Into<Vec<u8>> {
+        let l = (rand::random::<u8>() % (u8::MAX - 1)) + 1;
+        let mut v = Vec::with_capacity(l.into());
+        for _ in 0..l + 1 {
```
the increment is not needed here, as you already make sure that `l > 0` will always be true.

felsweg-iota commented on client/src/procedures/primitives.rs:
```diff
@@ -0,0 +1,1008 @@
+}
+
+// TODO: remove notes, add proper docs.
+#[derive(Clone, GuardDebug, Serialize, Deserialize)]
```
```suggestion
#[derive(Clone, GuardDebug, Serialize, Deserialize)]
#[non_exhaustive]
```

To future-proof the APi, better add `#[non_exhaustive]` to the enum, so added variants might be handled with a proper error.

felsweg-iota commented on client/src/procedures/primitives.rs:
```diff
@@ -0,0 +1,1008 @@
+
+        let mnemonic = bip39::wordlist::encode(
+            &entropy,
+            &bip39::wordlist::ENGLISH, // TODO: make this user configurable
```
we should somehow tackle this.

elenaf9 replied:
I think we should revisit the `BIP39Generate` Procedure in general. At the moment, the generated mnemonic is only used to derive the seed, an then just thrown away. So technically the language doesn't matter, as it is never exposed to the outside, but at the same time this procedure isn't useful for the user, because they have no way to know the mnemonic. wallet.rs for example doesn't use this procedure, but instead generates the mnemonic outside of stronghold and then uses `Bip39Recover` to create the seed record in stronghold.
I would change the `BIP39Generate` procedure to 1. make the wordlist configurable, and 2. return the mnemonic to the user. What do you think @felsweg-iota?

felsweg-iota replied:
I agree. The wordlist can be passed as parameter but as enum, and the resulting mnemonic should be returned to the user.

felsweg-iota commented on client/src/procedures/primitives.rs:
```diff
@@ -0,0 +1,1008 @@
+        }
+    }
+
+    pub fn dynamic(msg_key: OutputKey, key: Location) -> Self {
```
the functions `new` and `dynamic` (with similar intend) appear more than once. Might it be possible to combine them into a trait?

elenaf9 replied:
Generally, I think is a good suggestion and would be desirable to have such a trait. The problem is that the different procedures have different number of input arguments, so adding a trait would require in some cases to wrap the other input parameters  in a new struct, so we can have the signatures:
```
fn new(input: T, others: U) -> Self;
fn dynamic(msg_key: OutputKey, other: U) -> Self
```
with `T` being the input type, and `U`  being other parameters.
I'm afraid that this might make it more complicated for the user when they want to use these functions.

felsweg-iota commented on client/src/procedures/primitives.rs:
```diff
@@ -0,0 +1,1008 @@
+
+impl<T> AeadDecrypt<T> {
+    pub fn new(
+        key: Location,
```
maybe fewer arguments to the function call? Might there be a possibility to combine them into a parameters struct, that can be reused among individual calls?

tensor-programming replied:
I do kind of agree here, it would clean up the code a little to make a decryption input type or something along those lines.

felsweg-iota commented on client/src/state/secure.rs:
```diff
@@ -133,6 +138,119 @@ impl SecureClient {
     }
 }
 
+impl Runner for SecureClient {
```
:+1: for combining relevant function calls into here

felsweg-iota reviewed (changes requested):
Looks good to me.

elenaf9 committed:
feat(procs): rename procedure traits

Rename traits for primitive Procedures:
- Parse -> ProcessOutput
- Generate -> GenerateSecret
- Process -> DeriveSecret
- Utilize -> UseSecret

elenaf9 committed:
fix(procs): change BIP39Generate; apply review

elenaf9 committed:
feat(procs): Add error types

elenaf9 committed:
fix(procs): rename ProcessOutput -> ProcessData

elenaf9 committed:
feat(procs): better naming, clean code

elenaf9 committed:
feat(procs): use fully qualified syntax in macro

elenaf9 committed:
doc(procs): add code docs

tensor-programming commented on client/src/state/secure.rs:
```diff
@@ -74,32 +79,32 @@ impl SecureClient {
 
     /// Resolves a location to a `VaultId` and a `RecordId`
-    pub fn resolve_location<L: AsRef<Location>>(&self, l: L) -> (VaultId, RecordId) {
+    pub fn resolve_location<L: AsRef<Location>>(l: L) -> (VaultId, RecordId) {
```
Since `resolve_location`, `derive_vault_id` and `derive_record_id`  no longer need to be methods, might be worth thinking about making them into functions for the Location API or something.   Calling `Self::` on these doesn't make sense when the input/output and logic don't really have anything to do with the `SecureClient`.   Not worth exposing them publicly from the client library but would clean up the code to at least just remove them from the impl block since they are effectively just wrapper functions.

felsweg-iota replied:
This will be fixed with another PR.

tensor-programming reviewed (approved):
Overall looks really nice and well thought out.  Will prove very useful for adding new procs in the future.

elenaf9 committed:
Merge branch 'dev' of github.com:iotaledger/stronghold.rs into feat/procedure-api

elenaf9 committed:
feat(client): clean re-exports

elenaf9 committed:
feat(client): add x25519 key and diffie-hellman

elenaf9 committed:
fix(procs): handle exec_proc when vid0 == vid1

elenaf9 committed:
fix(procs): fix & test Aead

elenaf9 committed:
test(procs): test x25519 diffie-hellman

elenaf9 committed:
test(procs): test bip39 recover

elenaf9 committed:
feat(procs): add pbkdf2

elenaf9 committed:
feat(procs): configure algorithm types via enums

elenaf9 committed:
feat(procs): add hkdf

elenaf9 committed:
feat(client): create vault on write if not found

felsweg-iota commented on client/src/actors/secure.rs:
```diff
@@ -625,388 +318,10 @@ impl_handler!(
-            }
-        }
+    fn handle(&mut self, proc: Procedure, _: &mut Self::Context) -> Self::Result {
+        proc.run(self)
```
:heart:

felsweg-iota reviewed (approved):
Awesome PR! We now have a (cryptographic) procedures API :partying_face:

elenaf9 committed:
feat(procs): clean types, add docs

elenaf9 committed:
feat(procs): implement default for TempProduct
